### PR TITLE
sdk: enforce that `Keypair::from_bytes` bytes match secret-derived `Pubkey`

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

`ed25519_dalek` v1 `Keypair` construction [footgun](https://rustsec.org/advisories/RUSTSEC-2022-0093.html)

#### Summary of Changes

fail `solana_sdk::signer::keypair::Keypair::from_bytes` if the specified `Pubkey` does not match the one derived from the specified `SecretKey`

this will backstop our wrapper types from the overly permissive underlying api

### CODE REVIEW

**Review by joncinque — APPROVED** [review]
Looks great! (Once CI passes)

File: sdk/src/signer/keypair.rs [hunk 1]
```diff
@@ -155,7 +164,7 @@ impl EncodableKeypair for Keypair {
 /// Reads a JSON-encoded `Keypair` from a `Reader` implementor
 pub fn read_keypair<R: Read>(reader: &mut R) -> Result<Keypair, Box<dyn error::Error>> {
     let bytes: Vec<u8> = serde_json::from_reader(reader)?;
-    let dalek_keypair = ed25519_dalek::Keypair::from_bytes(&bytes)
+    let dalek_keypair = Keypair::from_bytes(&bytes)
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply sdk/src/signer/keypair.rs]: yeah already fixed

File: sdk/src/signer/keypair.rs [hunk 2]
```diff
@@ -41,7 +41,16 @@ impl Keypair {
...
-        ed25519_dalek::Keypair::from_bytes(bytes).map(Self)
+        let secret =
+            ed25519_dalek::SecretKey::from_bytes(&bytes[..ed25519_dalek::SECRET_KEY_LENGTH])?;
+        let public =
+            ed25519_dalek::PublicKey::from_bytes(&bytes[ed25519_dalek::SECRET_KEY_LENGTH..])?;
+        let expected_public = ed25519_dalek::PublicKey::from(&secret);
+        (public == expected_public)
+            .then_some(Self(ed25519_dalek::Keypair { secret, public }))
+            .ok_or(ed25519_dalek::SignatureError::from_source(String::from(
+                "keypair bytes do not specify same pubkey as derived from secret their key",
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply sdk/src/signer/keypair.rs]: thanks, fixed with ci

**Comment by t-nelson** [comment]
lmao... we have illegal keypair bytes in a test :upside_down_face:

**Comment by t-nelson** [comment]
> lmao... we have illegal keypair bytes in a test 🙃

looks like this key material has existed [since we used the `ring` crate for ed25519](https://github.com/solana-labs/solana/blob/d96a6b42a59ea76f5b2334fafbb91de0bf2f2612/sdk/src/transaction.rs#L290). i'm guessing we botched converting the key from pkcs8 format when porting over to dalek :shrug:

**Comment by t-nelson** [comment]
> i'm guessing we botched converting the key from pkcs8 format when porting over to dalek 🤷

confirmed. [this patch](https://github.com/solana-labs/solana/commit/a9cfae486c9dfe4af84a3919723c7eb3b68ee497#diff-8bf829cafa4d83c28318511be76cdf37b5e00c7c363778c312aef103114f6b2cR286) should've been something like...

```diff
@@ -1,5 +1,4 @@
-        48, 83, 2, 1, 1, 48, 5, 6, 3, 43, 101, 112, 4, 34, 4, 32, 255, 101, 36, 24, 124, 23,
-        167, 21, 132, 204, 155, 5, 185, 58, 121, 75, 156, 227, 116, 193, 215, 38, 142, 22, 8,
-        14, 229, 239, 119, 93, 5, 218, 161, 35, 3, 33, 0, 36, 100, 158, 252, 33, 161, 97, 185,
-        62, 89, 99, 195, 250, 249, 187, 189, 171, 118, 241, 90, 248, 14, 68, 219, 231, 62, 157,
-        5, 142, 27, 210, 117,
+        255, 101, 36, 24, 124, 23, 167, 21, 132, 204, 155, 5, 185, 58, 121, 75, 156, 227, 116,
+        193, 215, 38, 142, 22, 8, 14, 229, 239, 119, 93, 5, 218, 36, 100, 158, 252, 33, 161, 97,
+        185, 62, 89, 99, 195, 250, 249, 187, 189, 171, 118, 241, 90, 248, 14, 68, 219, 231, 62,
+        157, 5, 142, 27, 210, 117,
```

original bytes break down as...
```bash
        # some pkcs8 bullshit
        48, 83, 2, 1, 1, 48, 5, 6, 3, 43, 101, 112, 4, 34, 4, 32,
        # raw private key
        255, 101, 36, 24, 124, 23, 167, 21, 132, 204, 155, 5, 185, 58, 121, 75, 156, 227, 116, 193, 215, 38, 142, 22, 8, 14, 229, 239, 119, 93, 5, 218,
        # more pkcs8 bullshit
        161, 35, 3, 33, 0,
        # raw public key
        36, 100, 158, 252, 33, 161, 97, 185, 62, 89, 99, 195, 250, 249, 187, 189, 171, 118, 241, 90, 248, 14, 68, 219, 231, 62, 157, 5, 142, 27, 210, 117,
```

will correct in followup

**Pull request merged**